### PR TITLE
As/ft to forger pool

### PR DESCRIPTION
**Pull request by i-skrypnyk**
## Description
Adds a WellKnownAddress - Forger reward pool, that is used to send fund to a forger reward pool. Funds are distributed during block forging.
Adds support for next operations:
- FT to a forger reward pool
- FT to a native smart contract
- FT to a deployed smart contract
- transaction to a forger reward pool

*note - from a SC perspective, FT and coinbase transactions are identical

## Jira Ticket
https://horizenlabs.atlassian.net/browse/SDK-1365
https://horizenlabs.atlassian.net/browse/SDK-1364

## Checks
- [x] Project Builds
- [x] Project passes tests and checks
- [x] Updated documentation accordingly
- [x] Breaking changes have been correctly tagged and notified

### CODE REVIEW

**Review by saratnt — CHANGES_REQUESTED** [review]
This new feature seems to require an hard fork but in this PR the support for it is currently missing. Is it suppose to be done in a future task?

File: sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala [hunk 1]
```diff
@@ -122,20 +123,30 @@ class StateDbAccountStateView(
...
           )
 
-          if (isEoaAccount(recipientProposition.address())) {
+          val recipientAddress = recipientProposition.address()
+
+          if (recipientAddress.equals(WellKnownAddresses.FORGER_POOL_RECIPIENT_ADDRESS)) {
+            // move funds to forger pool
+            log.debug(s"adding FT amount = $value to forger pool")
+            mcForwardTransfersToForgerPoolAmount = mcForwardTransfersToForgerPoolAmount.add(value)
+          } else if (isEoaAccount(recipientAddress) && canReceiveFunds(recipientAddress)) {
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala:132]: we supported FT to EOA before, so I think we should keep doing that. The problem is, all of the native smart contracts are also return true for the condition isEoaAccount, so I had to filter them out somehow here also

File: sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala [hunk 2]
```diff
@@ -370,4 +381,15 @@ class StateDbAccountStateView(
 
   override def getGasTrackedView(gas: GasPool): BaseAccountStateView =
     new StateDbAccountStateViewGasTracked(stateDb, messageProcessors, gas)
+
+  /**
+   * Checks if the given address is not one of the well known addresses that cannot receive funds.
+   */
+  private def canReceiveFunds(address: Address): Boolean = {
+    !Seq(
+      WellKnownAddresses.FORGER_STAKE_SMART_CONTRACT_ADDRESS,
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala:390]: Added it, thanks!

File: sdk/src/main/scala/io/horizen/account/state/AccountState.scala [hunk 3]
```diff
@@ -127,20 +127,21 @@ class AccountState(
...
+      var cumGasUsed: BigInteger = BigInteger.ZERO
+      var cumBaseFee: BigInteger = BigInteger.ZERO // cumulative base-fee, burned in eth, goes to forgers pool
+      var cumForgerTips: BigInteger = BigInteger.ZERO // cumulative max-priority-fee, is paid to block forger
+
       for (mcBlockRefData <- mod.mainchainBlockReferencesData) {
         stateView.addTopQualityCertificates(mcBlockRefData, mod.id)
-        stateView.applyMainchainBlockReferenceData(mcBlockRefData)
+        val mcForwardTransfersToForgerPoolAmount = stateView.applyMainchainBlockReferenceData(mcBlockRefData)
+        cumBaseFee = cumBaseFee.add(mcForwardTransfersToForgerPoolAmount)
       }
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/AccountState.scala:139]: The implementation was reworked, now there is no calculations during the block forging, all the extra fee distribution is done as part of usual fee distribution process. More details below.

File: sdk/src/main/scala/io/horizen/account/forger/AccountForgeMessageBuilder.scala [hunk 4]
```diff
@@ -96,20 +96,28 @@ class AccountForgeMessageBuilder(
...
     for (mcBlockRefData <- mainchainBlockReferencesData) {
       // Since forger still doesn't know the candidate block id we may pass random one.
       val dummyBlockId: ModifierId = bytesToId(new Array[Byte](32))
       stateView.addTopQualityCertificates(mcBlockRefData, dummyBlockId)
-      stateView.applyMainchainBlockReferenceData(mcBlockRefData)
+      val mcForwardTransfersToForgerPoolAmount = stateView.applyMainchainBlockReferenceData(mcBlockRefData)
+      cumBaseFee = cumBaseFee.add(mcForwardTransfersToForgerPoolAmount)
+    }
+
+    if (stateView.getBalance(WellKnownAddresses.FORGER_POOL_RECIPIENT_ADDRESS).signum() == 1) {
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/forger/AccountForgeMessageBuilder.scala:111]: The implementation was reworked, now there is no calculations during the block forging, all the extra fee distribution is done as part of usual fee distribution process. More details below.

File: qa/sc_evm_forging_fee_payments.py [hunk 5]
```diff
@@ -187,6 +187,15 @@ def run_test(self):
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply qa/sc_evm_forging_fee_payments.py:187]: Yes, added that as part of the new test, thanks!

**Comment by i-skrypnyk** [comment]
Solution reworked after the new requirements added.
Changes:
- added fork, extra fee distribution is protected by a fork
- nodes keep track of number of forged blocks per forger to be able to calculate extra forger fee correctly
- extra fee distribution only happens at the end of withdrawal epoch
- extra fee distribution is part of regular fee distribution calculation

**Review by saratnt — CHANGES_REQUESTED** [review]

File: sdk/src/test/scala/io/horizen/account/utils/AccountFeePaymentsUtilsTest.scala [hunk 6]
```diff
@@ -153,4 +153,50 @@ class AccountFeePaymentsUtilsTest
...
+    val abfi_c1 = AccountBlockFeeInfo(
+      baseFee = BigInteger.valueOf(100),
+      forgerTips = BigInteger.valueOf(10),
+      forgerAddr_c)
+    val abfi_c2 = AccountBlockFeeInfo(
+      baseFee = BigInteger.valueOf(100),
+      forgerTips = BigInteger.valueOf(10),
+      forgerAddr_c)
+
+    val mcForgerPoolRewards = Map(
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/test/scala/io/horizen/account/utils/AccountFeePaymentsUtilsTest.scala:180]: added condition into the tests, thanks

File: sdk/src/main/scala/io/horizen/account/storage/AccountStateMetadataStorageView.scala [hunk 7]
```diff
@@ -399,6 +409,40 @@ class AccountStateMetadataStorageView(storage: Storage) extends AccountStateMeta
...
+          case Failure(e) =>
+            log.error("Failed to parse forger block counters from storage", e)
+            Map.empty[AddressProposition, Long]
+        }
+      case _ => Map.empty[AddressProposition, Long]
+    }
+  }
+
+  def resetForgerBlockCounters(): Unit = {
+    forgerBlockCountersOpt = Some(Map.empty[AddressProposition, Long])
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/storage/AccountStateMetadataStorageView.scala:436]: The whole process was reworked, as discussed on a call

File: sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala [hunk 8]
```diff
@@ -378,4 +382,16 @@ class StateDbAccountStateView(
...
+      WellKnownAddresses.WITHDRAWAL_REQ_SMART_CONTRACT_ADDRESS,
+      WellKnownAddresses.CERTIFICATE_KEY_ROTATION_SMART_CONTRACT_ADDRESS,
+      WellKnownAddresses.MC_ADDR_OWNERSHIP_SMART_CONTRACT_ADDRESS
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala]: The condition is removed, as this list also removed

File: qa/sc_evm_ft_to_smart_contract.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+Tests:
+    - Test 1 - FT to a native Smart Contract address
+    - Test 2 - FT to a deployed Smart Contract address
+    - Test 3 - negative - FT to a native Smart Contract address that cannot receive funds,
+               check that the burn address is updated
+"""
+
+
+class SCEvmFtToNativeContract(AccountChainSetup):
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply qa/sc_evm_ft_to_smart_contract.py:26]: Added tests before and after the fork, thanks!

File: sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala [hunk 10]
```diff
@@ -108,18 +108,23 @@ class StateDbAccountStateView(
             MainchainTxCrosschainOutputAddressUtil.getAccountAddress(ftOut.propositionBytes)
           )
 
-          if (isEoaAccount(recipientProposition.address())) {
+          val recipientAddress = recipientProposition.address()
+
+          if (isEoaAccount(recipientAddress) && canReceiveFunds(recipientAddress)) {
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/StateDbAccountStateView.scala]: Added the check, thanks!

File: qa/sc_evm_forging_fee_payments.py
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply qa/sc_evm_forging_fee_payments.py:1]: Now combined forger pool fee test into this one, and it should pass

File: sdk/src/main/scala/io/horizen/account/state/AccountState.scala [hunk 11]
```diff
@@ -379,7 +385,35 @@ class AccountState(
 
   override def getFeePaymentsInfo(withdrawalEpoch: Int, blockToAppendFeeInfo: Option[AccountBlockFeeInfo] = None): Seq[AccountPayment] = {
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/AccountState.scala]: The process is reworked as discussed on a call

File: sdk/src/main/scala/io/horizen/account/utils/AccountFeePaymentsUtils.scala [hunk 12]
```diff
@@ -17,7 +18,7 @@ object AccountFeePaymentsUtils {
     }
   }
 
-  def getForgersRewards(blockFeeInfoSeq : Seq[AccountBlockFeeInfo]): Seq[AccountPayment] = {
+  def getForgersRewards(blockFeeInfoSeq : Seq[AccountBlockFeeInfo], mcForgerPoolRewards: Map[AddressProposition, BigInteger] = Map.empty): Seq[AccountPayment] = {
     if (blockFeeInfoSeq.isEmpty)
       return Seq()
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/utils/AccountFeePaymentsUtils.scala]: Changed this condition, so that if blockFeeInfoSeq is empty, we just take mcForgerPoolRewards and turn them into AccountPayment. If mcForgerPoolRewards is also empty, this would return empty Seq

File: sdk/src/main/scala/io/horizen/account/state/AccountState.scala [hunk 13]
```diff
@@ -377,9 +385,11 @@ class AccountState(
 
   override def hasCeased: Boolean = stateMetadataStorage.hasCeased
 
-  override def getFeePaymentsInfo(withdrawalEpoch: Int, blockToAppendFeeInfo: Option[AccountBlockFeeInfo] = None): Seq[AccountPayment] = {
+  override def getFeePaymentsInfo(withdrawalEpoch: Int, consensusEpochNumber: ConsensusEpochNumber, blockToAppendFeeInfo: Option[AccountBlockFeeInfo] = None): Seq[AccountPayment] = {
```
Comment (saratnt) [inline sdk/src/main/scala/io/horizen/account/state/AccountState.scala:388]: This method is called in AbstractSidechainNodeViewHolder after having applied the block for saving the fees inside the History. The problem is that the rewards from the Forger_Poll have already been canceled, so at this point they are always 0. Maybe these fees need to be saved also in the metadatastorage

File: sdk/src/test/scala/io/horizen/account/utils/AccountFeePaymentsUtilsTest.scala [hunk 14]
```diff
@@ -153,4 +155,54 @@ class AccountFeePaymentsUtilsTest
...
+    )
+
+
+    blockFeeInfoSeq = blockFeeInfoSeq :+ abfi_a
+    blockFeeInfoSeq = blockFeeInfoSeq :+ abfi_b
+    blockFeeInfoSeq = blockFeeInfoSeq :+ abfi_c1
+    blockFeeInfoSeq = blockFeeInfoSeq :+ abfi_c2
+
+    val accountPaymentsList = getForgersRewards(blockFeeInfoSeq, mcForgerPoolRewards)
+    assertEquals(accountPaymentsList.length, 4)
```
Comment (saratnt) [inline sdk/src/test/scala/io/horizen/account/utils/AccountFeePaymentsUtilsTest.scala:195]: The parameters order should be inverted, because in `assertEqual` the first parameter is the expected value and the second one is the actual one. This applies also to other `assertEqual` in this test.

File: qa/sc_evm_ft_to_smart_contract.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        # Repeat the FTs, now they should work
+        self.advance_to_epoch(60)
+
+        # TEST 2.1 - ft to native SC - success
+        self.check_ft_to_smart_contract(FORGER_STAKE_SMART_CONTRACT_ADDRESS, True)
+
+        # TEST 2.2 - ft to deployed SC - success
+        self.check_ft_to_smart_contract(smart_contract_address, True)
+
+        # TEST 1.3 - FT to a random address - success
```
Comment (saratnt) [inline qa/sc_evm_ft_to_smart_contract.py:109]: Typo: Test 1.3 => Test 2.3 (I think)

File: qa/sc_evm_ft_to_smart_contract.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    NULL_ADDRESS, FORGER_STAKE_SMART_CONTRACT_ADDRESS
+from SidechainTestFramework.scutil import generate_next_blocks, generate_next_block, EVM_APP_SLOT_TIME, SLOTS_IN_EPOCH
+from test_framework.util import (assert_equal, forward_transfer_to_sidechain, assert_false)
+
+"""
+Configuration: 
+    - 1 SC node
+    - 1 MC node
+
+Tests:
```
Comment (saratnt) [inline qa/sc_evm_ft_to_smart_contract.py:18]: This list of tests needs to be updated.

**Review by saratnt — CHANGES_REQUESTED** [review]

File: qa/sc_evm_forging_fee_payments.py [hunk 17]
```diff
@@ -285,6 +316,49 @@ def run_test(self):
...
+        sc_node_2_balance_after_payments = sc_node_2.wallet_getTotalBalance()['result']['balance']
+        assert_equal(sc_node_1_balance_before_payments + node_1_fees, sc_node_1_balance_after_payments,
+                     "Wrong fee payment amount for SC node 1")
+        assert_equal(sc_node_2_balance_before_payments + node_2_fees, sc_node_2_balance_after_payments,
+                     "Wrong fee payment amount for SC node 2")
+
+        # assert forger pool balance is 0 now, as the fees are distributed
+        forger_pool_balance = int(self.sc_nodes[0].rpc_eth_getBalance(format_evm(FORGER_POOL_RECIPIENT_ADDRESS), 'latest')['result'], 16)
+        assert_equal(forger_pool_balance, 0)
+
```
Comment (paolocappelletti) [inline qa/sc_evm_forging_fee_payments.py:361]: we need to test (here or in a separate test) also this corner case please:
- withdrawal epoch N ends and fees are redistributed
- a revert on mainchain happens, causing the sidechain to go back to previous withdrawal epoch N-1
- balances are reverted correctly
- new  blocks are forgerd and we switch again withrawal epoch N
- fees are redistributed again correctly

**Review by paolocappelletti — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/io/horizen/AbstractSidechainNodeViewHolder.scala [hunk 18]
```diff
@@ -442,7 +442,8 @@ abstract class AbstractSidechainNodeViewHolder[
 
               val stateWithdrawalEpochNumber: Int = stateAfterApply.getWithdrawalEpochInfo.epoch
               val (historyResult, walletResult) = if (stateAfterApply.isWithdrawalEpochLastIndex) {
-                val feePayments : FPI = getFeePaymentsInfo(stateAfterApply, stateWithdrawalEpochNumber)
+                val consensusEpochNumber: ConsensusEpochNumber = stateAfterApply.getCurrentConsensusEpochInfo._2.epoch
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/AbstractSidechainNodeViewHolder.scala:445]: There's no `getConsensusEpochNumber` in AbstractState unfortunately, so I have to take it like this

File: sdk/src/main/scala/io/horizen/account/state/ForgerBlockCountersSerializer.scala [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    w.putInt(forgerBlockCounters.size)
+    forgerBlockCounters.foreach { case (address, counter) =>
+      addressSerializer.serialize(address, w)
+      w.putLong(counter)
+    }
+  }
+
+  override def parse(r: Reader): Map[AddressProposition, Long] = {
+    val length = r.getInt()
+    (1 to length).map { _ =>
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply sdk/src/main/scala/io/horizen/account/state/ForgerBlockCountersSerializer.scala:21]: yes, added a test for that case

**Review by saratnt — CHANGES_REQUESTED** [review]
For me it's OK now, I have just very minor comments

**Review by saratnt — APPROVED** [review]

**Review by paolocappelletti — APPROVED** [review]

File: sdk/src/main/scala/io/horizen/AbstractSidechainNodeViewHolder.scala [hunk 20]
```diff
@@ -442,7 +442,8 @@ abstract class AbstractSidechainNodeViewHolder[
...
+                val feePayments : FPI = getFeePaymentsInfo(stateAfterApply, stateWithdrawalEpochNumber, consensusEpochNumber)
```
Comment (alsala) [inline sdk/src/main/scala/io/horizen/AbstractSidechainNodeViewHolder.scala:446]: Consider getting `consensusEpochNumber` the same way but inside the implementation of the `getFeePaymentsInfo()` method using the state passed as input parameter. In this way we can avoid modifying the signature of this abstract method (not all overridden methods really uses that value)

File: sdk/src/main/scala/io/horizen/account/forger/AccountForgeMessageBuilder.scala [hunk 21]
```diff
@@ -267,11 +259,13 @@ class AccountForgeMessageBuilder(
               val withdrawalEpochNumber: Int = WithdrawalEpochUtils.getWithdrawalEpochInfo(mainchainBlockReferencesData.size, dummyView.getWithdrawalEpochInfo, params).epoch
 
               // get all previous payments for current ending epoch and append the one of the current block
-              val feePayments = dummyView.getFeePaymentsInfo(withdrawalEpochNumber, Some(currentBlockPayments))
+              val feePayments = dummyView.getFeePaymentsInfo(withdrawalEpochNumber, consensusEpochNumber, Some(currentBlockPayments))
```
Comment (alsala) [inline sdk/src/main/scala/io/horizen/account/forger/AccountForgeMessageBuilder.scala:262]: I am not sure, but couldn't we use the `consensusEpochNumber` of the `dummyView` instance instead of passing the one got from the blockContext? Should not they be the same?

Inside the AccountStateView class, something like:

    `val consensusEpochNumber: ConsensusEpochNumber = getConsensusEpochNumber.getOrElse(intToConsensusEpochNumber(0))`

Comment (alsala) [reply sdk/src/main/scala/io/horizen/account/forger/AccountForgeMessageBuilder.scala:262]: Ok I was wrong, the reason is that if we are on the edge of end-epoch the next block can result in having the next epoch value
